File: src/assignment1/src/transformation_node.cpp
#include "ros/ros.h"
#include "assignment1/NewPosition.h"
#include <math.h>

#define PI 3.14159265

bool transform(assignment1::NewPosition::Request &req, assignment1::NewPosition::Response &res){
  float a = cos(req.gamma*PI/180);
  float b = sin(req.gamma*PI/180);
  float c = cos(req.beta*PI/180);
  float d = sin(req.beta*PI/180);
  float e = cos(req.alpha*PI/180);
  float f = sin(req.alpha*PI/180);
  
  res.xnew = (c*a*req.x) - (req.y*e*b) + (req.y*f*d*a) + (req.z*f*b) + (req.z*e*d*a);
  res.ynew = c*b*req.x - (req.z*f*a) + req.y*e*a + req.y*f*d*b + req.z*e*d*b;
  res.znew = -d*req.x + req.y*f*c + req.z*e*c;

  res.xnew = res.xnew + req.d;
  return true;
}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "transformation");
  ros::NodeHandle nh;

  ros::ServiceServer service = nh.advertiseService("Transformation", transform);
  
  ros::spin();
  return 0;
}

File: src/assignment1/scripts/transformation_node.py
import rospy
import numpy as np
from assignment1.srv import NewPosition, NewPositionResponse
import math

def callback(request):
    alph = math.pi/180 * request.alpha
    bet = math.pi/180 * request.beta
    gam = math.pi/180 * request.gamma
    xx = np.array([[1,0,0], [0, math.cos(alph), math.sin(-1*alph)], [0, math.sin(alph), math.cos(alph)]])
    yy = np.array([[math.cos(bet),0,math.sin(bet)], [0,1,0], [math.sin(-1*bet), 0, math.cos(bet)]])
    zz = np.array([[math.cos(gam), math.sin(-1*gam), 0], [math.sin(gam), math.cos(gam),0], [0,0,1]])
    result = np.dot(np.dot(np.dot(xx, yy), zz), np.array([[request.x], [request.y], [request.z]]))
    return NewPositionResponse(result[0]+request.d, result[1], result[2])

def transform():
    rospy.init_node("Transformation")
    service = rospy.Service("transform", NewPosition, callback)
    rospy.spin()

if __name__ == '__main__':
    transform()